### ISSUE TITLE
[Bug]: Trains on different track types can collide in end-to-end stations

### ISSUE THREAD

**Issue by DFEvans** [OpenTTD/OpenTTD#10493]
### Version of OpenTTD

13.0, Windows 10

### Expected result

Trains of different types (railway, monorail, maglev) arrive in the end-to-end stations without colliding (as neither will pass onto the 'other' track type), or the signalling system will prevent them from using the track simultaneously.

### Actual result

The trains are allowed to use the station simultaneously and collide.

### Steps to reproduce

To recreate from scratch:
1. Place a Railway, Monorail, or Maglev station
2. Using a different track type, place another station on the end of the first, as if you were extending it
3. Timetable trains to use both halves of the station, and ensure they will be stopped in the platform simultaneously
4. When the second of the two trains arrives, it will collide with the one already in the platform

Attached save has this scenario set up, with a Monorail - Maglev station. After unpausing, the monorail train will arrive second, colliding with the maglev.

[MonoMaglev.zip](https://github.com/OpenTTD/OpenTTD/files/10776589/MonoMaglev.zip)

Station layout in question - a 1x3 Monorail station end-on to a 1x3 Maglev station.
![image](https://user-images.githubusercontent.com/28623784/219942015-ee6ae0d2-8337-4291-b422-515d810cc5a3.png)

(Based on a report on Reddit: https://www.reddit.com/r/openttd/comments/115mc5d/in_rare_cases_using_a_path_signal_can_cause/)

**Comment by 2TallTyler** [comment]
Related issue, the same can happen with depots: #8424

**Comment by PeterN** [comment]
Yeah, the cause is trains stop when the middle of the front engine hits the end, rather than the front of the front end.

**Comment by PeterN** [comment]
Retracted that comment, the trains do not actually go past the end of the tile in this instance, but they do touch.

**Comment by 2TallTyler** [comment]
Fixed by #14366.
